File: frontend/workbench/app/PlanProcessOrder.js
(function () {
  'use strict';
  const object = value => value !== null && typeof value === 'object' && !Array.isArray(value);
  const exact = (value, keys) => object(value) && Reflect.ownKeys(value).length === keys.length && keys.every(key => Object.prototype.hasOwnProperty.call(value, key));
  const ref = value => typeof value === 'string' && /^[0-9a-f]{48}$/.test(value);
  function validate(value, data) {
    if (!exact(value, ['state', 'basis', 'items', 'issues']) || !Array.isArray(value.items) || value.items.length > 10000
        || !Array.isArray(value.issues) || !value.issues.every(row => exact(row, ['code', 'message']) && typeof row.code === 'string' && row.code && typeof row.message === 'string' && row.message)) return false;
    if (value.state === 'unavailable') return value.basis === null && !value.items.length && value.issues.length > 0;
    if (value.state !== 'available' || !['run_admission', 'trial_creation'].includes(value.basis) || value.issues.length) return false;
    const tasks = new Map(), operations = new Map();
    for (const item of value.items) {
      if (!exact(item, ['task_ref', 'operation_ref', 'predecessor_operation_refs']) || !ref(item.task_ref) || !ref(item.operation_ref) || tasks.has(item.task_ref)
          || !Array.isArray(item.predecessor_operation_refs) || item.predecessor_operation_refs.length > 10000
          || !item.predecessor_operation_refs.every(ref) || new Set(item.predecessor_operation_refs).size !== item.predecessor_operation_refs.length
          || item.predecessor_operation_refs.includes(item.operation_ref)) return false;
      tasks.set(item.task_ref, item.operation_ref);
      const previous = JSON.stringify(item.predecessor_operation_refs.slice().sort());
      if (operations.has(item.operation_ref) && operations.get(item.operation_ref) !== previous) return false;
      operations.set(item.operation_ref, previous);
    }
    if (!data.tasks.every(task => tasks.get(task.task_ref) === task.operation_ref)
        || data.scope.range_start === null && tasks.size !== data.tasks.length) return false;
    const incoming = new Map(), next = new Map();
    for (const [operation, encoded] of operations) {
      const parents = JSON.parse(encoded).filter(parent => operations.has(parent));
      incoming.set(operation, parents.length);
      for (const parent of parents) { if (!next.has(parent)) next.set(parent, []); next.get(parent).push(operation); }
    }
    const queue = Array.from(incoming).filter(([, count]) => count === 0).map(([operation]) => operation);
    for (let index = 0; index < queue.length; index++) for (const child of next.get(queue[index]) || []) {
      const count = incoming.get(child) - 1; incoming.set(child, count); if (!count) queue.push(child);
    }
    return queue.length === operations.size;
  }
  function relationships(data, selected) {
    const order = data.projections.process_order, task = selected && selected.task;
    if (!task || selected.before || order.state !== 'available') return null;
    const current = order.items.find(row => row.task_ref === task.task_ref);
    if (!current) return null;
    const tasks = new Map(data.tasks.map(row => [row.task_ref, row])), byOperation = new Map();
    for (const row of order.items) {
      if (!byOperation.has(row.operation_ref)) byOperation.set(row.operation_ref, []);
      byOperation.get(row.operation_ref).push(row);
    }
    const expand = operations => operations.flatMap(operation => {
      const rows = byOperation.get(operation);
      return rows ? rows.map(row => ({ task_ref: row.task_ref, task: tasks.get(row.task_ref) || null })) : [{ task_ref: null, task: null }];
    });
    const successors = new Set(order.items.filter(row => row.predecessor_operation_refs.includes(task.operation_ref)).map(row => row.operation_ref));
    return { previous: expand(current.predecessor_operation_refs), next: expand(Array.from(successors)) };
  }
  window.PlanProcessOrder = { validate, relationships };
})();

File: core/services/workbench/plan_process_order.py
"""Frozen process relations for one complete adopted plan, never current BOM."""

from types import SimpleNamespace

from core.models.workbench_command import WorkbenchCommandRejected
from core.models.workbench_piece_adoption import PieceAdoptionBlocked
from core.models.workbench_plan_reference import WorkbenchPlanReferenceError
from data.repositories.workbench_plan_identity_repo import WorkbenchPlanIdentityRepository

from .piece_adoption_scope import build_piece_adoption_scope
from .plan_adoption_baseline_identity import _live_sources
from .plan_adoption_baseline_values import AdoptionBaselineUnavailable, indexed, require
from .plan_projection import check_payload_size, read_adopted_source
from .trial_adoption_storage import load_saved_scenario


def _unavailable(facts, code, gap):
    message = ("此计划未记录工序顺序。" if code == "process_order_not_recorded" else
               "工序顺序数据不完整。")
    facts.update(outcome=code, evidence_gap=gap)
    return {"state": "unavailable", "basis": None, "items": [],
            "issues": [{"code": code, "message": message}]}, facts


def _frozen_operations(conn, tables, arranged):
    operations = indexed(tables["BatchOperations"], "id")
    batches = indexed(tables["Batches"], "batch_id")
    wanted = {row["op_id"] for row in arranged}
    require(wanted <= set(operations), "process_order.original_operations")
    batch_ids = {operations[key]["batch_id"] for key in wanted}
    require(batch_ids <= set(batches), "process_order.original_batches")
    related = {key: row for key, row in operations.items() if row["batch_id"] in batch_ids}
    keys = {str(key) for key in related}
    sources = [row for row in tables["WorkbenchPlanSourceRefs"]
               if row["kind"] == "operation" and row["active"] == 1 and row["source_key"] in keys]
    by_key = indexed(sources, "source_key")
    require(set(by_key) == keys, "process_order.complete_operation_refs")
    _live_sources(conn, sources)
    refs = {key: by_key[str(key)]["ref"] for key in related}
    require(len(set(refs.values())) == len(refs), "process_order.unique_operation_refs")
    return related, {key: batches[key] for key in batch_ids}, refs


def _candidate_relations(operations, batches, refs):
    scope = build_piece_adoption_scope([SimpleNamespace(**row) for row in operations.values()],
                                      {key: SimpleNamespace(**row) for key, row in batches.items()})
    return {refs[row.op_id]: [refs[key] for key in row.predecessor_op_ids] for row in scope.operations}


def _trial_relations(conn, audit, refs):
    _, _, rows = load_saved_scenario(conn, audit["scenario_ref"])
    known = set(refs.values())
    relations = {}
    for row in rows:
        original = row["original"]
        op_id = original["operation"]["id"]
        require(op_id in refs and row["operation_ref"] == refs[op_id], "process_order.trial_operation_identity")
        previous = original["predecessor_operation_refs"]
        require(type(previous) is list and len(set(previous)) == len(previous) and set(previous) <= known
                and refs[op_id] not in previous, "process_order.trial_frozen_predecessors")
        require(row["operation_ref"] not in relations, "process_order.trial_duplicate_operation")
        relations[row["operation_ref"]] = previous
    return relations


def _acyclic(relations):
    counts = {key: 0 for key in relations}
    following = {key: [] for key in relations}
    for key, parents in relations.items():
        for parent in parents:
            if parent in counts:
                counts[key] += 1
                following[parent].append(key)
    ready = [key for key, count in counts.items() if not count]
    index = 0
    while index < len(ready):
        for child in following[ready[index]]:
            counts[child] -= 1
            if not counts[child]:
                ready.append(child)
        index += 1
    require(len(ready) == len(relations), "process_order.frozen_cycle")


def _project(conn, plan_ref, rows, relations, refs):
    repository = WorkbenchPlanIdentityRepository(conn)
    tasks = repository.get_task_refs(plan_ref, [dict(row, schedule_id=row["id"]) for row in rows])
    all_refs = {row[0] for row in conn.execute("SELECT ref FROM WorkbenchTaskRefs WHERE plan_ref=?", (plan_ref,))}
    require(all_refs == set(tasks.values()) and len(tasks) == len(rows), "process_order.complete_task_membership")
    items = []
    for row in rows:
        operation_ref = refs[row["op_id"]]
        require(operation_ref in relations, "process_order.missing_frozen_relation")
        items.append({"task_ref": tasks[row["id"]], "operation_ref": operation_ref,
                      "predecessor_operation_refs": list(relations[operation_ref])})
    require(len({item["task_ref"] for item in items}) == len(items), "process_order.unique_task_refs")
    return items


def project_process_order(conn, *, plan_ref):
    """Return a complete read-only projection and the facts used by its fingerprint."""
    if not conn.in_transaction:
        raise RuntimeError("Process order reads require a caller-owned read transaction.")
    facts = {"plan_ref": plan_ref}
    try:
        evidence = read_adopted_source(conn, plan_ref)
        if evidence is None:
            return _unavailable(facts, "process_order_not_recorded", "no_audited_frozen_source")
        basis, audit, tables, arranged, selected = evidence
        facts.update(adoption=audit, selected_rows=selected)
        operations, batches, refs = _frozen_operations(conn, tables, arranged)
        if basis == "candidate_adoption":
            relations = _candidate_relations(operations, batches, refs)
        else:
            relations = _trial_relations(conn, audit, refs)
        _acyclic(relations)
        items = _project(conn, plan_ref, selected, relations, refs)
        data = {"state": "available", "basis": "run_admission" if basis == "candidate_adoption" else "trial_creation",
                "items": items, "issues": []}
        check_payload_size(data)
        facts.update(relations=relations, items=items)
        return data, facts
    except AdoptionBaselineUnavailable as exc:
        return _unavailable(facts, "process_order_unavailable", exc.code + ":" + exc.gap)
    except (WorkbenchPlanReferenceError, PieceAdoptionBlocked) as exc:
        return _unavailable(facts, "process_order_unavailable", exc.code)
    except WorkbenchCommandRejected as exc:
        if exc.status == 413:
            raise
        return _unavailable(facts, "process_order_unavailable", exc.code)
    except (KeyError, TypeError, ValueError, OverflowError) as exc:
        return _unavailable(facts, "process_order_unavailable", "frozen_shape:" + type(exc).__name__)
